File: CPP/create_library_with_header/publisher_library/src/publisher_library.cpp
/**
 * @file publisher_library.cpp
 *
 * @brief A basic publisher library made with header file
 *
 * @author Antonio Mauro Galiano
 * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
 *
 */

#include "publisher_library/publisher_library.h"

PublisherLibrary::PublisherLibrary() : Node("ros2_publisher_library"), counter_(0)
{
  pub_ = this->create_publisher<std_msgs::msg::String>("/my_published_msg", 10);
  timer_ = this->create_wall_timer(
    std::chrono::milliseconds(500), std::bind(&PublisherLibrary::TimerCallback, this));
}

void PublisherLibrary::TimerCallback()
{
  std_msgs::msg::String message;
  message.data = "HELLO WORLD number " + std::to_string(counter_++);
  RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
  pub_->publish(message);
}

PublisherLibrary::~PublisherLibrary() {}

File: PYTHON/create_library_with_header_py/publisher_library_py/publisher_library_py/publisher_library.py
#  *
#  * @file publisher_library.py
#  *
#  * @brief A basic publisher library made with Python module
#  *
#  * @author Antonio Mauro Galiano
#  * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
#  *
#

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class PublisherLibrary(Node):
    """Publisher library class that publishes messages periodically."""

    def __init__(self):
        super().__init__('ros2_publisher_library')
        self.counter_ = 0
        self.pub_ = self.create_publisher(String, '/my_published_msg', 10)
        self.timer_ = self.create_timer(0.5, self.timer_callback)

    def timer_callback(self):
        """Timer callback that publishes a message."""
        message = String()
        message.data = f'HELLO WORLD number {self.counter_}'
        self.get_logger().info(f"Publishing: '{message.data}'")
        self.pub_.publish(message)
        self.counter_ += 1

